GPS with heading and proper redundancy

oystub opened the pull request:
Based on initial testing of https://github.com/aviant-tech/PX4-Autopilot/pull/104, update the GNSS selection and fallback.

The most important change is to automatically return to GNSS heading operation after a non-heading fallback. 
Not in scope: How to notify the pilot of the current redundancy state of the drone.

The existing cluttered blending logic is replaced with a shorter GNSS selector module, to make it clear that we follow exactly the drafted spec with clear fallback conditions:

# TODO
- [x] Tests
- [x] Point to correct driver submodule commit
- [x] Handling single-GNSS drones in a good way
- [x] Evaluate if we need separate body-frame offsets for each GPS 
   - [x] Should be ok for less than 50 cm (as it's only wrong during fallback). We have thousands of historic flights with Here3 before we set correct offset with larger errors. 

# Background spec
_NB! This is also documented on Confluence. This description might not be up to date, if you read this into the future._

GNSS receivers are the sole source of the drone’s XY position and ground-relative velocity in the local tangential (NED) frame. To ensure reliability, the system employs two redundant u-blox F9P receivers. Redundancy primarily protects against individual receiver or antenna faults.

Both receivers are multiband (L1 + L2), which provides robustness against multipath effects and RF interference compared to single-band operation.

In addition to position and velocity, the receiver pair provides a heading measurement derived from differential Real-Time Kinematic (RTK) carrier-phase measurements between the two units. This is the primary heading reference for the drone.

## 1. Primary/Secondary Roles

### A. Primary:
- U-blox F9P configured in Moving Base mode. 
- Always preferred as the selected GNSS source, provided it passes health checks.

### B. Secondary:
- U-blox F9P configured in Moving Base Rover mode, in which it computes a NED-frame heading relative to the primary unit.
- Rover mode does not provide reliable position data in the absence of the moving base.
- If secondary becomes the selected source, it must be reconfigured into Normal mode to ensure reliable operation.
- Reconfiguration between Rover and Normal mode requires 1-2 seconds.

## 2. Health Criteria

A GNSS unit is considered healthy if both A and B conditions below are fulfilled.

### A. Fix requirement

- Receiver reports 3D fix or better.

### B. Message deadline requirement

- The receiver has sent a message to the autopilot within the last 0.33 seconds.

## 3. Selection Rules

### A. Default Priority:

- The primary receiver is always preferred if healthy.

### B. Switching to Secondary:

- If the primary receiver becomes unhealthy, secondary is automatically selected.
  - Note that secondary health isn’t checked, since it might not fulfill the deadline requirement while in rover mode.
- Upon selection, if the rover is in Rover mode, it is reconfigured into Normal mode.

### C. Switching Back to Primary:
- If the secondary is selected and the primary regains health, switching back occurs only after a hysteresis period of 5 seconds (primary must remain healthy for 5 seconds).
- Exception: If the secondary is unhealthy, while primary is healthy, the switch to primary occurs immediately (no hysteresis).

## 4: Reconfiguration Rules for Secondary

### A. Normal Mode Activation:
- Mandatory whenever the secondary is selected.
- If not already in Normal mode, the receiver is reconfigured when switching.

### B. Heading Mode Restoration:
- If secondary has not been selected for hysteresis period of 10 seconds, it may be reconfigured back to Rover/Heading mode.
- Reconfiguration should not occur prematurely, to avoid rapid mode switching.

## 5. Summary of Hysteresis Conditions
### Selection Hysteresis (5 sec):
- Prevents immediate switching back to primary after recovery, unless the secondary fails.
### Reconfiguration Hysteresis (10 sec):
- Prevents premature reconfiguration of secondary back to heading mode after primary is selected.

Torjusba commented on src/drivers/gps/params.c:
```diff
@@ -253,17 +249,15 @@ PARAM_DEFINE_INT32(GPS_2_GNSS, 0);
- *
- * Seconds after boot before PX4 may reconfigure a u-blox F9P rover out of Rover With Moving Base.
- * Prevents premature resets and gives the receiver time to initialize.
+ * u-blox MB rover position reconfiguration hysterisis
```
```suggestion
 * u-blox MB rover position reconfiguration hysteresis
```

oystub replied:
Will fix this typo everywhere

Torjusba commented on src/drivers/gps/gps.cpp:
```diff
@@ -957,30 +960,46 @@ GPS::run()
+					}
+
+					if (used_for_position) {
+						_last_used_for_position = hrt_absolute_time();
```
It's a bit misleading to have _last_used_for_position=0 in the driver for the MB gps, since it is actually used for position.
We should either make it clear that the variable applies only when rover or have this variable updated for both MB and Rover, then check for rover-ness right right before the reconfiguration, e.g. if (ubx_mode == [rover] && used_for_position && _disabled_rover_mode) { _helper->disableUbxMBRover }

I think the second solution is a bit more readable, since we can avoid some nested conditionals that way

oystub replied:
Yes. Also found other issues here. Will fix

Torjusba commented on src/drivers/gps/gps.cpp:
```diff
@@ -957,30 +960,46 @@ GPS::run()
+						}
+
+					} else {
+						if (hrt_absolute_time() - _last_used_for_position > ubx_mbr_fb_hysterisis_us) {
```
Should also check `_disabled_rover_mode` here to avoid repeated re-configuring?

oystub replied:
Same as above

Torjusba commented on src/modules/sensors/vehicle_gps_position/gnss_selector.cpp:
```diff
@@ -0,0 +1,83 @@
+	const bool primary_ok   = _receiver_state[_primary_instance].is_healthy;
+	const bool secondary_ok = _receiver_state[secondary_instance].is_healthy;
+
+	if (!_initialized) {
```
`_initialized` is never set to true

oystub replied:
Yes, this should also be caught by tests

Torjusba reviewed (changes requested):
I really like the simplicity of the logic, makes it super easy to verify.
Left some comments and suggestions, let me know if you want to discuss anything.

Also, should we publish the status of the primary/secondary GNSS in some way? To enable preflight checks requiring two healthy GNSS moduels for verified redundancy.

Torjusba reviewed (commented):
I think the code looks good now, ready for testflight as long as sw and ground tests pass

Torjusba commented on src/modules/sensors/vehicle_gps_position/gnss_selector.hpp:
```diff
@@ -14,23 +20,35 @@ class GnssSelector
+			PX4_ERR("Invalid primary instance %d, must be between 0 and %d", instance, GNSS_MAX_RECEIVERS - 1);
+		}
+
+		_primary_instance = instance;
```
Should we just send the error message on invalid primary instance or also block the assignment? This could deserve a decision comment

oystub replied:
Fixed

Torjusba commented on src/modules/sensors/vehicle_gps_position/gnss_selector.hpp:
```diff
@@ -0,0 +1,75 @@
+	 */
+	bool isNewOutputDataAvailable() const { return _is_new_output_data_available; }
+
+	const sensor_gps_s &getOutputGnssData() const { return _receiver_state[_selected].gnss_data; }
```
Should probably verify that `_selected` is not -1 here, since this is public? To avoid negative indexing

oystub replied:
Fixed

Torjusba reviewed (approved):
lgtm for flight testing

oystub committed:
gps: reconfigure rover back to rover operation after fallback

With this change, boot-time based reconfiguration delay is no longer needed,
and has been removed

oystub committed:
vehicle_gps_position: nuke blending, use simpler selector instead

This makes it way easier to show that our GNSS fallback logic matches
our compliance documentation.
